File: 114N_regional_proper_event_summary.R
######
# Adam Griffin, 2020-04-27
#
# Summarising size and time, and extracting daily PoE of extreme events.
# 
# For aquaCAT, Project 07441.
# 
# Created ABG 2020-04-27
# Pipeline version ABG 2020-09-07
#
#
# OUTPUTS: eventdf_***.csv: dataframe of each event summarised. One event per row.
#
#####

if(interactive()){commandArgs <- function(...){c("01","present", "NW")}}

#### SETUP ####----------------------
if(substr(osVersion,1,3) == "Win"){
  source("S:/CodeABG/setup_script_00.R")
}else if (substr(osVersion,1,3) %in% c("Cen","Fed")){
  source("/prj/aquacat/CodeABG/setup_script_00.R")
}else{
  source("~/AQUACAT/CodeABG/setup_script_00.R")
}

if(settings$HTsumm){
  stop("Regional summaries exist. Stopping 114N.")
}

regions <- c("ANG", "ESC", "NE", "NSC", "NW", "SE", "SEV", "SSC", "SW",
             "THA", "TRE", "WAL")
REGIONS <- c("NW")

rn <- as.data.frame(rn)
r1 <- which(rn_regions$REGION == REGIONS[1])

nclustevent <- function(v, radii=50) {
    # This function estimates the number of "seperate clusters" of event going on during
    # a widespread event. 
    ne0 <- sum(v, na.rm=T)
    event1 <- rbind(as.matrix(rn[r1,][v,]), matrix(c(
      700, 0, 700000, 1000000
    ), nrow = 1))
    d <- dist(event1[, 1:2])
    d2 <- dist(event1[1:ne0, 1:2])
    if ((max(d2) > radii) & (ne0 >= 10)) {
      hh <- hclust(d, method = "single")
      ch <- max(cutree(hh, h = 0.8 * sort(hh$height, decreasing = T)[2])) - 1
      if(sort(hh$height,decreasing=T)[2] < 0.2*max(hh$height)){
        ch <- 1
      }
      ch <- min(ch, 4)
    } else{
      ch <- 1
    }
    ch
  }
  
peakyfun <- function(v, thr) {
  mn <- min(v, na.rm=T)
  if (mn > thr | mn > 1e8) {
    return(0)
  }else{
    mn50 <- mean(c(mn, thr))
    return(sum(v < mn50) / sum(v < thr))
  }
}

# thresh1 <- "POT2"
# ws1 <- "pc01"
# # Only using POT2 and 2% inundation minimums.
# jT <- which(threshName==thresh1)
# jW <- which(wsName == ws1)
print(paste("Running for threshold", thresh1, "at", ws1, "minimum spread."))

eventSummaryLine <- function(obs_events, i, threshold, r1=1:19914){
  ni <- eventNo[i]
  vvec <- sum(
    (ncvar_get(obs_events, "flow", start=c(1,i), count=c(-1,1)) > threshold),
    na.rm=T)
  avec <- ncvar_get(obs_events, "ape", start=c(1,i), count=c(-1,1))
  min_dvec <- min(
    ncvar_get(obs_events, "dpe", start=c(1,i), count=c(-1,1)),
    na.rm=T)
  D <- eventDayList[[jT]][[jW]][ni]
  L <- eventLList[[jT]][[jW]][ni]
  ncl <- nclustevent(avec < (1-exp(-2)))
  pk <- peakyfun(avec, (1-exp(-2)))

  return(list(i, D, L, vvec, min(avec), min_dvec, season(D), ncl, pk))
}

threshMat <- readRDS(paste0(data_wd, subfold, "threshMat_RCM",
                            RCM, suffix, ".rds"))
load(paste0(data_wd, subfold, "eventLists_RCM", RCM, suffix, ".RDa"))

param_table <- readdf(paste0(data_wd, subfold, "paramtableG_POT2_RCM",
                            RCM, suffix, ".csv"))


# for(REG in c("NW")){
REG <- "NW"
  ##### REGIONAL OBS SUMMARY #####-----------------------------------------------
  
r1 <- which(rn_regions$REGION == REG)

param_table <- param_table[r1,]
  
obs_events <- nc_open(paste0(data_wd,subfold, REG, "/",
                    "eventOBS_region_", REG, "_RCM", RCM, suffix, ".nc"))
  
NN <- obs_events$dim$event$len
  
eventSummaryFrame <- data.frame(eventNumber=numeric(NN),
                                  eventDay=numeric(NN),
                                  eventLength = numeric(NN),
                                  area = numeric(NN),
                                  peakA = numeric(NN),
                                  peakD = numeric(NN),
                                  season = character(NN),
                                  nclusters = numeric(NN),
                                  peakyness = numeric(NN),
                                  stringsAsFactors = FALSE)
  
  # (D %/% 90) %% 4 for season 0,1,2,3: DJF, MAM, JJA, SON
season <- function(D) switch(((D %/% 90) %% 4) + 1,
                             "DJF", "MAM", "JJA", "SON")
ST <- Sys.time()
ST0 <- Sys.time()

eventNo <- ncvar_get(obs_events, "eventNo")

for(i in 1:NN){
    if((i < 10) | (i %% 40 == 0)){ # time recording
    print(i)
    I <- difftime(Sys.time(), ST, units="secs")
    I0 <- difftime(Sys.time(), ST0, units="secs")
    print(paste("Percent remaining", 100*round((NN-i)/NN ,2)))
    print(paste("Time remaining", round((NN-i)/i * I0,2)))
    print(paste("Since last readout:", round(I,2)))
    ST <- Sys.time()
    print(ST) 
    }
  try({
  eventSummaryFrame[i,] <- eventSummaryLine(obs_events, i, threshMat[r1,jT])
  })
}

readr::write_csv(eventSummaryFrame, 
                 paste0(data_wd,subfold, REG, "/eventSumm_OBS_",REG,"_",
                        thresh1,"_", ws1,"_RCM", RCM, suffix, ".csv"))

print(paste("OBS regional summary done."))














### EC SUMMARY ####-------------------------------------------------------------
obs_events <- nc_open(paste0(data_wd,subfold, "/", REG, "/",
                  "eventEC_region_", REG, "_RCM", RCM, suffix, ".nc"))

NN <- obs_events$dim$event$len
eventSummaryFrame <- data.frame(eventNumber=numeric(NN),
                                eventDay=numeric(NN),
                                eventLength = numeric(NN),
                                area = numeric(NN),
                                peakA = numeric(NN),
                                peakD = numeric(NN),
                                season = character(NN),
                                nclusters = numeric(NN),
                                peakyness = numeric(NN),
                                stringsAsFactors = FALSE)

# (D %/% 90) %% 4 for season 0,1,2,3: DJF, MAM, JJA, SON
season <- function(D) switch(((D %/% 90) %% 4) + 1, "DJF", "MAM", "JJA", "SON")
r1 <- which(rn_regions$REGION == REG)

ST <- Sys.time()
ST0 <- Sys.time()
for(i in 1:NN){
  if((i < 10) | (i %% 40 == 0)){ # time recording
    print(i)
    I <- difftime(Sys.time(), ST, units="secs")
    I0 <- difftime(Sys.time(), ST0, units="secs")
    print(paste("Percent remaining", 100*round((NN-i)/NN ,2)))
    print(paste("Time remaining", round((NN-i)/i * I0,2)))
    print(paste("Since last readout:", round(I,2)))
    ST <- Sys.time()
    print(ST) 
  }
  #print(i)
  try({
  eventSummaryFrame[i,] <- eventSummaryLine(obs_events, i, threshMat[r1,jT])
  })
}

readr::write_csv(eventSummaryFrame, paste0(data_wd,subfold, REG,
                "/eventSumm_EC_",REG,"_", thresh1,"_", ws1,
                "_RCM", RCM, suffix, ".csv"))

print(paste("EC regional summary done."))










#### HT SUMMARY ####------------------------------------------------------------
if(settings$HTflow){
  
  obs_events <- nc_open(paste0(data_wd,subfold, REG, "/eventHT_",
                        REG,"_",thresh1,"_", ws1,"_RCM", RCM, suffix, ".nc"))
  
  NN <- obs_events$dim$event$len
  eventSummaryFrame <- data.frame(eventNumber=numeric(NN),
                                  eventDay=numeric(NN),
                                  eventLength = numeric(NN),
                                  area = numeric(NN),
                                  peakA = numeric(NN),
                                  peakD = numeric(NN),
                                  season = character(NN),
                                  nclusters = numeric(NN),
                                  peakyness = numeric(NN),
                                  stringsAsFactors = FALSE)
  
  # (D %/% 90) %% 4 for season 0,1,2,3: DJF, MAM, JJA, SON
  season <- function(D) switch(((D %/% 90) %% 4) + 1, "DJF", "MAM", "JJA", "SON")
  
  ST <- Sys.time()
  ST0 <- Sys.time()
  for(i in 1:NN){
    if((i < 10) | (i %% 40 == 0)){ # time recording
      print(i)
      I <- difftime(Sys.time(), ST, units="secs")
      I0 <- difftime(Sys.time(), ST0, units="secs")
      print(paste("Percent remaining", 100*round((NN-i)/NN ,2)))
      print(paste("Time remaining", round((NN-i)/i * I0,2)))
      print(paste("Since last readout:", round(I,2)))
      ST <- Sys.time()
      print(ST) 
      }
    #print(i)
    try({
    eventSummaryFrame[i,] <- eventSummaryLine(obs_events, i, threshMat[r1,jT])
    })
  }
  readr::write_csv(eventSummaryFrame,
                   paste0(data_wd,subfold, REG, "/eventSumm_HT_",REG,"_",
                          thresh1,"_", ws1,"_RCM", RCM, suffix, ".csv"))
  print("HT regional summary done.")
}
print(paste(REG, "region done."))


settings$HTsumm <- TRUE
write_yaml(settings, settingspath)

print("114N done.")

File: 114N_regional_proper_event_summary.py
# -*- coding: utf-8 -*-
"""
Created on Thu Sep  2 11:17:32 2021
Author: Adam Griffin, UKCEH
Project: AQUACAT
Script to summarise EC events.
"""
import os
import netCDF4 as nc
import pandas as pd
import numpy as np
import re
import sys
import yaml

rcm = sys.argv[1]

if sys.argv[2] == "present":
    period = "198012_201011"
else:
    period = "205012_208011"

print("Running RCM {} for {}.".format(rcm, period))

RCMS = ["01","04","05","06","07","08","09","10","11","12","13","15"]
periods = ["198012_201011","205012_208011"]

# CHANGE THIS TO THE TOP LEVEL OF THE FOLDER THE CSVs ARE IN
toplevel = r"/prj/aquacat/Data"

# CHANGE THIS TO THE TOP LEVEL OF THE FOLDER THE NETCDFs ARE IN
outlevel = toplevel #r'S:/Data'

# CHANGE THIS TO WHERE THE hasData files are, they should exist in the toplevel folder.
rn = pd.read_csv("".join([toplevel,"/hasData_primary.csv"]))
rnreg = pd.read_csv("".join([toplevel,"/hasData_Regions.csv"]))

regional = True

if regional:
    subfold='/NW'
    fileinfix = 'NW_POT2_pc01'
    rn = rn[rnreg.REGION=="NW"]
    NH = 1437
else:
    subfold=''
    fileinfix = 'POT2_pc01'
    NH = 19914

#for method in ["OBS", "EC", "HT"]:
method="OBS" # "EC" OR "OBS", "HT" AS NEEDED

ncpath = (f"{outlevel}/RCM{rcm}_{period}{subfold}/event{method}_{fileinfix}"
          f"_RCM{rcm}_{period}.nc")
ncfile = nc.Dataset(ncpath, mode='r')

param_path = (f"{outlevel}/RCM{rcm}_{period}{subfold}/paramtableG"
              f"_POT2_RCM{rcm}_{period}.csv")

param_table = pd.read_csv(param_path)

thresh_path = f"{outlevel}/RCM{rcm}_{period}{subfold}/threshMat_RCM{rcm}_{period}.csv"

threshvec = pd.read_csv(thresh_path).iloc[:,1]
if regional:
    threshvec = threshvec[rnreg.REGION == "NW"]

init_path = (f"{outlevel}/RCM{rcm}_{period}/initialSummary_RCM{rcm}_{period}.csv")
init_table = pd.read_csv(init_path)

summ_path = (f"{outlevel}/RCM{rcm}_{period}{subfold}/eventSumm"
             f"_OBS_POT2_pc01_RCM{rcm}_{period}.csv")


threshvec = pd.read_csv(thresh_path).iloc[:,1]
if regional:
    threshvec = threshvec[rnreg.REGION == "NW"]

summtable_out = pd.DataFrame(columns=["eventNumber", "eventDay", "eventLength",
                                      "area","peakA", "peakD", "season",
                                      "nclusters","peakyness"])

eventNo = list(ncfile.variables["eventNo"][:])
NE = np.sum([i > 0 for i in eventNo])

avec_all = ncfile.variables['ape'][:,:]
avec_mid = ncfile.variables['ape_mid'][:,:]
vvec_all = ncfile.variables['flow'][:,:]
dvec_all = ncfile.variables['dpe'][:,:]

ncfile.close()

print("Setup complete")

import time

start_time = time.time()
for i in range(NE):
    if (i < 10) or (i % 200) == 0:
        print(i)
        print("--- %s seconds ---" % (time.time() - start_time))
        start_time = time.time()
    ni = eventNo[i]
    vvec = sum(vvec_all[i,:] > threshvec)
    avec = min(avec_all[i,:])
    amid = min(avec_mid[i,:])
    dvec = min(dvec_all[i,:])
    D = summtable.iloc[ni-1,1:3]
    summtable_out.loc[i] = [ni, D[0], D[1], vvec, avec, dvec, summtable.iloc[ni-1,6], 0, 0]

print("--- %s seconds ---" % (time.time() - start_time))
print("done")

yaml_path = "".join([outlevel, '/RCM', rcm, '_', period, '/settings.yaml'])

summtable_out.to_csv("".join([outlevel, r'/RCM', rcm, '_', period, subfold,
                  r'/eventSumm_', method, 'B_', fileinfix,'_RCM',rcm,
                  '_',period, '.csv'])) 

with open(yaml_path) as ym:
    list_doc = yaml.safe_load(ym)

list_doc['HTsumm'] = True
list_doc['propsumm'] = "114N.py"

with open(yaml_path, 'w') as ym:
    yaml.dump(list_doc, ym)
    
    print(time.strftime("%Y-%m-%d %H:%M:%S"))
print("Files saved and YAML updated. End.")